File: src/levels/L25-un-solo-oyente/command.rb
# Command convierte la llamada en un objeto… que nadie encola, deshace ni registra.
class StartOrder
  def initialize(kitchen, order)
    @kitchen = kitchen
    @order = order
  end

  # region: StartOrder#call
  def call = @kitchen.start(@order)
  # endregion
end

class Cashier
  def initialize(kitchen) = @kitchen = kitchen

  # region: Cashier#charge
  def charge(order)
    CardTerminal.charge(order)
    StartOrder.new(@kitchen, order).call
  end
  # endregion
end

# region: wiring
OrderTaker.new(Cashier.new(Kitchen.new(Counter.new))).take(:latte)
# endregion

File: src/levels/L25-un-solo-oyente/ts/command.ts
// Command convierte la llamada en un objeto… que nadie encola, deshace ni registra.
class StartOrder {
  constructor(
    private kitchen: Kitchen,
    private order: Order,
  ) {}

  // region: StartOrder#call
  call() { this.kitchen.start(this.order) }
  // endregion
}

class Cashier {
  constructor(private kitchen: Kitchen) {}

  // region: Cashier#charge
  charge(order: Order) {
    CardTerminal.charge(order)
    new StartOrder(this.kitchen, order).call()
  }
  // endregion
}

// region: wiring
new OrderTaker(new Cashier(new Kitchen(new Counter()))).take('latte')
// endregion
